### PR TITLE
Feature/gdpr

### LINKED ISSUE #87: DSGVO command
The bot should be GDPR and DSGVO compatible.
Currently the user needs to request its data on discord. However this is a lot of manual work on dev side.

Goal is to implement a mechanic which allows user to:

- [x] Implement request of all data associated with a user.
- [x] Implement deletion of all data associated with a user.
- [ ] Implement command listener for private messages.

Data request should be done by writing a private message to a user. Implementation could be similar to the internal command listener.

Returned data should remove all data from other users. Data should be supplied as json dump of the requested database object.
Deletion of data should not destroy integrity of the database. Data will not be deleted, but anonymized. This means that user ids are purged and set to 0. No user should lose reputation caused by data deletion of another user.

### PR DESCRIPTION
**Pull request by RainbowDashLabs**
Adds a gdpr command to the bot which allows user to request their data or its deletion.

Deletion and sending of data will be done hourly.

Close #87

### CODE REVIEW

File: src/main/resources/database/1/patch_5.sql [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+             from repbot_schema.voice_activity
+         )
+    SELECT jsonb_build_object(
+                   'reputation', coalesce(rep, '[]'::jsonb),
+                   'voice_activity', coalesce(voice, '[]'::jsonb)
+               )
+    from reputation,
+         voice_activity
+    into _result;
+    return _result;
```
Comment (Taucher2003) [inline src/main/resources/database/1/patch_5.sql:44]:
```suggestion
    FROM reputation,
         voice_activity
    INTO _result;
    RETURN _result;
```

File: src/main/resources/database/1/patch_5.sql [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    user_id  bigint not null,
+    received timestamp
+);
+
+create unique index IF NOT EXISTS gdpr_log_user_id_uindex
+    on repbot_schema.gdpr_log (user_id);
+
+alter table repbot_schema.cleanup_schedule alter column guild_id set default 0;
+
+alter table repbot_schema.cleanup_schedule alter column user_id set default 0;
```
Comment (Taucher2003) [inline src/main/resources/database/1/patch_5.sql:59]:
```suggestion
CREATE TABLE IF NOT EXISTS repbot_schema.gdpr_log
(
    user_id  BIGINT NOT NULL,
    received TIMESTAMP
);

CREATE UNIQUE INDEX IF NOT EXISTS gdpr_log_user_id_uindex
    ON repbot_schema.gdpr_log (user_id);

ALTER TABLE repbot_schema.cleanup_schedule ALTER COLUMN guild_id SET DEFAULT 0;

ALTER TABLE repbot_schema.cleanup_schedule ALTER COLUMN user_id SET DEFAULT 0;
```

File: src/main/resources/locale.properties [hunk 3]
```diff
@@ -7,6 +7,13 @@ command.channel.sub.remove=Remove a reputation channel.
 command.channel.sub.remove.removed=Removed the following channel from reputation channel:\n%CHANNEL%
 command.channel.sub.set=Set the reputation channel
 command.channel.sub.set.set=Set the following channel as reputation channel:\n%CHANNEL%
+command.gdpr.description=Request a copy of data or its deletion.
+command.gdpr.sub.delete=Request deletion of your data.
+command.gdpr.sub.delete.received=Your data is scheduled to be deleted. It will be deleted within the next hours.
+command.gdpr.sub.delete.scheduled=Your data is already scheduled to be deleted. It will be deleted within the next hours.
+command.gdpr.sub.request=Request a copy or your data.
+command.gdpr.sub.request.received=We received your request. You will get your data within the next hours. You can request your data every 30 days.
+command.gdpr.sub.request.requested=Your have requested your data in the last 30 days. You can request your data only every 30 days.
```
Comment (Taucher2003) [inline src/main/resources/locale.properties:16]:
```suggestion
command.gdpr.sub.request.requested=You have requested your data in the last 30 days. You can request your data only every 30 days.
```

File: src/main/resources/locale_en_US.properties [hunk 4]
```diff
@@ -7,6 +7,13 @@ command.channel.sub.remove=Remove a reputation channel.
 command.channel.sub.remove.removed=Removed the following channel from reputation channel:\n%CHANNEL%
 command.channel.sub.set=Set the reputation channel
 command.channel.sub.set.set=Set the following channel as reputation channel:\n%CHANNEL%
+command.gdpr.description=Request a copy of data or its deletion.
+command.gdpr.sub.delete=Request deletion of your data.
+command.gdpr.sub.delete.received=Your data is scheduled to be deleted. It will be deleted within the next hours.
+command.gdpr.sub.delete.scheduled=Your data is already scheduled to be deleted. It will be deleted within the next hours.
+command.gdpr.sub.request=Request a copy or your data.
+command.gdpr.sub.request.received=We received your request. You will get your data within the next hours. You can request your data every 30 days.
+command.gdpr.sub.request.requested=Your have requested your data in the last 30 days. You can request your data only every 30 days.
```
Comment (Taucher2003) [inline src/main/resources/locale_en_US.properties:16]:
```suggestion
command.gdpr.sub.request.requested=You have requested your data in the last 30 days. You can request your data only every 30 days.
```

**Review by Taucher2003 — CHANGES_REQUESTED** [review]

**Review by DRSchlaubi — APPROVED** [review]

File: src/main/resources/locale_de_DE.properties [hunk 5]
```diff
@@ -7,6 +7,13 @@ command.channel.sub.remove=Entfernt einen Reputationskanal.
 command.channel.sub.remove.removed=Entfernt den folgenden Kanal von den Reputationskanälen:\n%CHANNEL%
 command.channel.sub.set=Setzen des Reputationskanals
 command.channel.sub.set.set=Setzt den folgenden Kanal als Reputationskanal:\n%CHANNEL%
+command.gdpr.description=Frage eine Kopie oder die Löschung deiner Daten an.
```
Comment (dlsf) [inline src/main/resources/locale_de_DE.properties:10]:
```suggestion
command.gdpr.description=Beantrage eine Kopie oder die Löschung deiner Daten.
```

File: src/main/resources/locale_de_DE.properties [hunk 6]
```diff
@@ -7,6 +7,13 @@ command.channel.sub.remove=Entfernt einen Reputationskanal.
...
+command.gdpr.sub.delete=Frage die Löschung deiner Daten an.
```
Comment (dlsf) [inline src/main/resources/locale_de_DE.properties:11]:
```suggestion
command.gdpr.sub.delete=Beantrage die Löschung deiner Daten.
```

File: src/main/resources/locale_de_DE.properties [hunk 7]
```diff
@@ -7,6 +7,13 @@ command.channel.sub.remove=Entfernt einen Reputationskanal.
...
+command.gdpr.sub.delete.received=Deine Daten sind zur Löschung freigegeben. Sie werden innterhalb der nächsten Stunden gelöscht.
```
Comment (dlsf) [inline src/main/resources/locale_de_DE.properties:12]:
```suggestion
command.gdpr.sub.delete.received=Deine Daten sind zur Löschung freigegeben. Sie werden innerhalb der nächsten Stunden entfernt.
```

File: src/main/resources/locale_de_DE.properties [hunk 8]
```diff
@@ -7,6 +7,13 @@ command.channel.sub.remove=Entfernt einen Reputationskanal.
...
+command.gdpr.sub.delete.scheduled=Deine Daten sind bereits zur Löschung freigegeben. Sie werden in den nächsten Stunden gelöscht.
```
Comment (dlsf) [inline src/main/resources/locale_de_DE.properties:13]:
```suggestion
command.gdpr.sub.delete.scheduled=Deine Daten sind bereits zur Löschung freigegeben. Sie werden in den nächsten Stunden entfernt.
```

File: src/main/resources/locale_de_DE.properties [hunk 9]
```diff
@@ -7,6 +7,13 @@ command.channel.sub.remove=Entfernt einen Reputationskanal.
...
+command.gdpr.sub.request=Frage eine Kopie deiner Daten an.
```
Comment (dlsf) [inline src/main/resources/locale_de_DE.properties:14]:
```suggestion
command.gdpr.sub.request=Beantrage eine Kopie deiner Daten.
```

**Review by dlsf — CHANGES_REQUESTED** [review]
There is a synonym for "beantragen" which is maybe more fitting but I forgot it

**Comment by RainbowDashLabs** [comment]
> There is a synonym for "beantragen" which is maybe more fitting but I forgot it

Nothing is for ever :D We can always change it in the future :D

**Review by dlsf — APPROVED** [review]

**Pull request merged**